#166: Add documentation to tapiro (closes #166) (#167)
buildo/retro: docs/tapiro/introduction.md

---
id: introduction
title: Introduction
---

Tapiro parses your Scala controllers to generate HTTP endpoints.

A Scala controller is a trait defined as follows:

```scala mdoc
import scala.annotation.StaticAnnotation

class query extends StaticAnnotation
class command extends StaticAnnotation

case class Cat(name: String)
case class Error(msg: String)

trait Cats[F[_], AuthToken] {
  @query //translate this to a GET
  def findCutestCat(): F[Either[Error, Cat]]

  @command //translate this to a POST
  def doSomethingWithTheCat(catId: Int, token: AuthToken): F[Either[Error, Unit]]
}
```

For each controller tapiro generates two files:
- `CatsEndpoints.scala` containing the HTTP api description using https://tapir-scala.readthedocs.io/
- `CatsHttp4sEndpoints.scala` or `CatsAkkaHttpEndpoints.scala` depeneding on the HTTP server the user is using.

The resulting routes can be simply used like this (http4s example):

> ```scala

Reviewer: this block is not typechecked

Reviewer: ☝️ farei un esempio più completo che effettivamente compili

Author: Per avere esempio completo che compili servono dipendenze, per questo stavo rimandando la cosa.

Author: Aggiunto superesempione.

Author: @gabro pingone

val routes = CatsHttp4sEndpoints.routes(catsImplementation)

override def run(args: List[String]): IO[ExitCode] =
  BlazeServerBuilder[IO]
    .bindHttp(port, host)
    .withHttpApp(routes.orNotFound)
    .serve
    .compile
    .drain
    .as(ExitCode.Success)
```

The resulting server can be queried as follows:
```
/GET /Cats/findCutestCat
/POST /Cats/doSomethingWithTheCat -d '{ "catId": 1 }'
```

## Authentication

An `AuthToken` type argument is expected in each controller and is added as authorization header.

`trait Cats[F[_], AuthToken]`

> The acutal implementation of the `AuthToken` is left to the user. All tapiro requires is a proper tapir `PlainCodec` such as:

Reviewer: typo: actual

- `CatsEndpoints.scala` containing the HTTP api description using https://tapir-scala.readthedocs.io/
- `CatsHttp4sEndpoints.scala` or `CatsAkkaHttpEndpoints.scala` depeneding on the HTTP server the user is using.

## Complete Example

> Here you have an example implementation of the `Cats` controller definied in the previous section:

Reviewer: ```suggestion
Here you have an example implementation of the `Cats` controller defined in the previous section:
```

```scala mdoc
import cats.effect._

class CatsImpl[F[_]](implicit F: Sync[F]) extends Cats[F, String] {
  override def findCutestCat(): F[Either[Error, Cat]] = F.delay(Right(Cat("Cheshire")))
  override def doSomethingWithTheCat(catId: Int, token: String): F[Either[Error, Unit]] = F.delay(Right(()))

> }

Reviewer: ```suggestion
object Cats {
  def create[F[_]](implicit F: Sync[F]) = new Cats[F, String] {
    override def findCutestCat(): F[Either[Error, Cat]] =
      F.delay(Right(Cat("Cheshire")))
    override def doSomethingWithTheCat(catId: Int, token: String): F[Either[Error, Unit]] =
      F.delay(Right(()))
  }
}
```

```

> Here you have the autogenerated magic fromo tapiro (This is the content of `CatsHttp4sEndpoints.scala` it will be autogenerated).

Reviewer: ```suggestion
Here you have the http4s endpoints auto-generated by tapiro (this would be the content of `CatsHttp4sEndpoints.scala`) 
```

```scala mdoc
import org.http4s.HttpRoutes

// ---- begins autogenerated code
object CatsHttp4sEndpoints {
  def routes(controller: Cats[IO, String]): HttpRoutes[IO] = ???
}
// ---- ends autogenerated code
```

Here is how to run the server:

```scala mdoc
import org.http4s.server.blaze._

> import org.http4s.implicits._

Reviewer: serve questo import?

Author: sì

Author: magari riprovo

Author: confermo che serve per `.orNotFound`

Reviewer: 🤦‍♂ ok

import cats.implicits._

object Main extends IOApp {

>   val catsImpl = new CatsImpl[IO]()

Reviewer: as per suggestion above:

```suggestion
  val catsImpl = CatsImpl.create[IO]
```

Author: ah già

import cats.implicits._

object Main extends IOApp {

>   val catsImpl = CatsImpl.create[IO]

Reviewer: ops, era `Cats.create[IO]`

```suggestion
  val catsImpl = Cats.create[IO]
```
